internal/env/baseconfig.go

```
@@ -8,6 +8,15 @@ import (
 	"github.com/hashicorp/go-multierror"
 )
 
+type Config interface {
+	// Load is called prior to env.Lock an application startup. This method should
+	// read the values from the environment and store errors to be reported later.
+	Load()
```

This is a general question: what's the rationale behind the signature being `Load()` instead of `Load() error`?

---

Since the config is meant to be used with our internal env package, we need something to call before and after `env.Lock()`. Load is meant to be the thing that registers it config values, and Validate is meant to ensure that everything was read properly.

In the case of the worker, it will allow us to load configuration values, and then only care about validation errors conditionally (if there's a missing config for a disabled process, we wouldn't care).

It also might make a little more sense when you see how it's used in practice (which makes for very terse code validating user environment). This is from the config of the precise-code-intel-worker. (Validate is defined by the nested BaseConfig struct).

```go
type Config struct {
	env.BaseConfig

	UploadStoreConfig  *uploadstore.Config
	WorkerPollInterval time.Duration
	WorkerConcurrency  int
	WorkerBudget       int64
}

func (c *Config) Load() {
	uploadStoreConfig := &uploadstore.Config{}
	uploadStoreConfig.Load()
	c.UploadStoreConfig = uploadStoreConfig

	c.WorkerPollInterval = c.GetInterval("PRECISE_CODE_INTEL_WORKER_POLL_INTERVAL", "1s", "Interval between queries to the upload queue.")
	c.WorkerConcurrency = c.GetInt("PRECISE_CODE_INTEL_WORKER_CONCURRENCY", "1", "The maximum number of indexes that can be processed concurrently.")
	c.WorkerBudget = int64(c.GetInt("PRECISE_CODE_INTEL_WORKER_BUDGET", "0", "The amount of compressed input data (in bytes) a worker can process concurrently. Zero acts as an infinite budget."))
}
```

---

Here, `c.GetInterval` will decode an interval from the environment and stash an error if the user value is not parseable.

---

I see - is the idea behind env.lock that we have multiple things registering configuration, and that we need to have a way of ensuring that we have a consistent/stable view on the environment once we start validating (and failing fast - panicking if someone tries to register configuration afterwards). 

If so that makes sense to me. Could we clarify the documentation for Validate (in this interface) to say that it does, infact, perform validation instead of just returning errors?

---

Updated interface docs.